godtools/App/Services/Renderer/Views/MobileContent/Views/MobileContentView.swift

```diff
@@ -144,23 +144,74 @@ class MobileContentView: UIView {
     // MARK: - Events
     
     func sendEventsToAllViews(eventIds: [EventId], rendererState: State) {
+            
+        var eventIds: [EventId] = eventIds.flatMap({$0.resolve(state: rendererState)})
+                
+        walkUpHierarchyAndSendEventsToChildren(view: self, eventIds: &eventIds)
+                
+        recurseChildrenAndSendEventsFromView(view: getRootView(), eventIds: &eventIds)
+    }
+    
+    private func walkUpHierarchyAndSendEventsToChildren(view: MobileContentView, eventIds: inout [EventId]) {
         
-        let eventIds: [EventId] = eventIds.flatMap({$0.resolve(state: rendererState)})
+        var nextParentView: MobileContentView? = view.parent
         
-        recurseChildrenAndSendEvents(view: getRootView(), eventIds: eventIds)
+        while nextParentView != nil && !eventIds.isEmpty {
+            
+            if let children = nextParentView?.children {
+                
+                for childView in children {
+                    
+                    sendEventsToView(view: childView, eventIds: &eventIds)
```

Comment:
hmm, events may be triggered twice because you trigger events on the walk up, but also from the root view

Reply:
imagine this xml:
```xml
<page>
  <content:paragraph>
    <content:animation toggleListener="eventA" />
    <content:button event="eventA" />
  </content:paragraph>
</page>
```

I believe walking up you will trigger the animation `toggleListener` because you process children of all ancestor nodes. Plus you will trigger `toggleListener` again when you start recursively iterating from the root node.

Reply:
you probably want a method signature like this for the walk up:

```
func handleSpecialEvents(eventId: EventId) -> Boolean {
     guard parent != null else {
        return false
     }
     
     return parent.handleSpecialEvents(eventId)
}
```

with the form view overriding it like this:

```
func handleSpecialEvents(eventId: EventId) -> Boolean {
   guard eventId == EventId.Companion().FOLLOW_UP else {
     return super(eventId)
   }
   
   // submit followup event for this form
   
   return true
}
```

then when triggering events you could do something like:
```
for eventId in eventIds {
  if (!view.handleSpecialEvents(eventId)) {
    recurseChildrenAndSendEvents(eventId)
  }
}
```

Reply:
Hey @frett I like that approach.

Reply:
Hey @frett I think this should work now.  I created a separate method signature ```didReceiveAncestryEventFromWalkingUpViewHierarchy``` that views can listen for instead of the plain ```didReceiveEvent```.  The ```ToolPageFormView``` will now look for the new ancestry event signature when processing those forms.  I decided to leave the remove eventId logic for added measure of removing that follow up eventId when processed.